python/lsst/daf/butler/core/config.py

```diff
@@ -217,10 +225,35 @@ def __initFromFile(self, path):
             To a persisted config file.
         """
         if path.endswith("yaml"):
-            self.__initFromYamlFile(path)
+            if path.startswith('s3://'):
+                self.__initFromS3File(path)
+            else:
+                self.__initFromYamlFile(path)
         else:
             raise RuntimeError("Unhandled config file type:%s" % path)
 
+    def __initFromS3File(self, path):
+        """Load a file from path pointing to an S3 Bucket.
+
+        Parameters
+        ----------
+        path : `str`
+            To a persisted config file.
+        """
+        if boto3 is None:
+            raise ModuleNotFoundError(('boto3 not found.'
+                                       'Are you sure it is installed?'))
+
+        uri = ButlerURI(path)
+        s3 = boto3.client('s3')
+        try:
+            response = s3.get_object(Bucket=uri.netloc, Key=uri.path.lstrip('/'))
```

Comment:
Because `boto3` (AWS S3 protocol itself really) does not handle keys as gracefully as paths are handled in POSIX. So `/path/` and `/path` is a different thing, even when we're pointing to the same directory and `/path` is interpreted as inserting two keys into the bucket, one empty and one hierarchically under that called `path` that will look like a file (while we wanted a directory). Or there's something I don't understand about URIs. I'm not sure, but this is what I have noticed.                 
At least algorithmically, S3 protocol seems to do something like:

```
key = '/some/path/file.ext'
for subkey in key.split('/'):
    s3.insert_key(subkey)
```
Since split will produce an empty string for leading and trailing separators '/' we will be inserting different keys into the Bucket. On a concrete example:

```
import boto3
s3 = boto3.client('s3')

s3.put_object(Bucket='lsst', Key='/there/is/a/difference', Body=b'testtest')
s3.put_object(Bucket='lsst', Key='/there/is/a/difference/', Body=b'testtest')

s3.put_object(Bucket='lsst', Key='there/is/a/difference', Body=b'testtest')
s3.put_object(Bucket='lsst', Key='there/is/a/difference/', Body=b'testtest')
```

will upload 4 different keys to which urls will be:

```
https://lsstspark.s3-us-west-2.amazonaws.com//there/is/a/difference
https://lsstspark.s3-us-west-2.amazonaws.com//there/is/a/difference//   *

https://lsstspark.s3-us-west-2.amazonaws.com/there/is/a/difference
https://lsstspark.s3-us-west-2.amazonaws.com/there/is/a/difference//    *
```
\* there would be no objects under that path, but the keys would exists. 

Notice the double-slashes in the URLs to the keys whenever there is a leading or trailing separator '/'. As far as the code goes, it would work just normally. But in practical terms the directory and file structure in the bucket would look completely off when compared to the local filesystem layout. 

If I want the bucket layout to match that of the local filesystem datastore directory layout, which I believe is sensible and practical) we need to be context sensitive. Is it a directory I am creating, is it a file I am writing, is it a directory that needs to be hierarchically nested under another directory or is it some just really deeply nested file that I am uploading are the different cases.

If we want to insert a key that acts as a directory, at the root level of the bucket, we need to insert `key` and not `/key` because the second one will insert two keys (effectively making two directories) one of which will have a blank name and the other one that will be called `key` under it. For a sub directory it's `key/subkey/` while for a file in `key` it's `key/subkey`.                    
Otherwise navigation through the directory structure in the bucket very awkward as there is a nameless directory between every hierarchical level. 

This does not seem completely unreasonable thing to me, as long as I imagine all paths are taken relative to bucket name (i.e. netloc). 

So keeping that in mind I have renamed the `bucketName` to `netloc` and `pathInBucket` to `relativeToNetloc`. Now, of course, there's the question of, current ButlerURI, implementation not having a `netloc` whenever local filesystem paths are being supplied as URIs\*\*. So to make that method mean what it's named as; I've changed the `ButlerURI` to use a `localhost` as a `netloc` whenever local filesystem is being used.  Note that they should still resolve correctly and are correct URIs since the translation:

```
file://localhost/rootDir/    --->    file:///rootDir/
file://localhost/rootDir    --->    file:///rootDir
```
is a given as netloc does not necessarily have to be specified as part of the hierarchical part of an uri. This also requires me to expand netloc and scheme for the paths for which we force root, or assume current directory is used

```
ButlerURI('relative/file.ext, root='tmpRoot', forceAbsolute=True) --> file://localhost/tmp/relative/file.ext
ButlerURI('relative/file.ext, forceAbsolute=False) --> relative/file.ext
ButlerURI('~/relative/file.ext', forceAbsolute=False) -->  file://localhost/home/user/relative/file.ext
```

Is this acceptable compromise?